# Task

## T5.1: Download the 'Portugal_online_retail', 'Sweden_online_retail, and 'UK_online_retail' datasets from Canvas. Apply the apriori algorithm to all datasets using three different confidence levels. Select one confidence level for each dataset that you think works better. Determine the first three most important rules for each dataset using the selected confidence level and report them in the report cell. Explain what each rule means (Completing the report cell is required) (15%).

NOTE: You should comment on your code

In [1]:
# import libraries and install necessary modules
import pandas as pd;pd.set_option('display.max_columns', None)
import numpy as np
!pip install mlxtend
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, fpmax, fpgrowth
import warnings;warnings.filterwarnings('ignore')

# For Portugal_Online_Retails

In [2]:
#import the dataset
portugal_basket = pd.read_csv('Portugal_online_retail.csv')

In [3]:
# View the dataset
display(portugal_basket.head())

InvoiceNo  10 COLOUR SPACEBOY PEN  12 PENCIL SMALL TUBE WOODLAND  \
0     536990                       0                              0   
1     537246                       0                              0   
2     537818                       0                              0   
3     537915                       0                              0   
4     538311                       0                              0   

   12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS TALL TUBE POSY  \
0                                    0                          0   
1                                    0                          0   
2                                    0                          0   
3                                    0                          0   
4                                    0                          0   

   12 PENCILS TALL TUBE RED RETROSPOT  12 PENCILS TALL TUBE WOODLAND  \
0                                   0                              0   
1                                   0                              0   
2                                   0                              0   
3                                   0                              0   
4                                   0                              0   

   20 DOLLY PEGS RETROSPOT  3 PIECE SPACEBOY COOKIE CUTTER SET  \
0                        0                                   0   
1                        0                                   0   
2                        0                                   0   
3                        0                                   0   
4                        0                                   0   

   3 STRIPEY MICE FELTCRAFT  36 FOIL HEART CAKE CASES  \
0                         0                         0   
1                         1                         0   
2                         0                         0   
3                         0                         0   
4                         0                         0   

   36 PENCILS TUBE WOODLAND  3D TRADITIONAL CHRISTMAS STICKERS  \
0                         0                                  0   
1                         0                                  0   
2                         0                                  0   
3                         0                                  0   
4                         0                                  0   

   4 TRADITIONAL SPINNING TOPS  50'S CHRISTMAS GIFT BAG LARGE  \
0                            0                              0   
1                            0                              0   
2                            0                              0   
3                            0                              0   
4                            0                              0   

   50'S CHRISTMAS PAPER GIFT BAG  6 GIFT TAGS 50'S CHRISTMAS  \
0                              0                           0   
1                              0                           0   
2                              0                           0   
3                              0                           0   
4                              0                           0   

   6 GIFT TAGS VINTAGE CHRISTMAS  60 CAKE CASES DOLLY GIRL DESIGN  \
0                              0                                0   
1                              0                                0   
2                              0                                0   
3                              0                                0   
4                              0                                0   

   60 TEATIME FAIRY CAKE CASES  72 SWEETHEART FAIRY CAKE CASES  \
0                            0                               0   
1                            0                               0   
2                            0                               0   
3                            0                               0   
4                            0                               0   

   

In [4]:
# dropping the InvoiceNo column
portugal_basket = portugal_basket.drop('InvoiceNo',axis=1)

In [5]:
#view new portugal data without invoice column
portugal_basket.head()

10 COLOUR SPACEBOY PEN  12 PENCIL SMALL TUBE WOODLAND  \
0                       0                              0   
1                       0                              0   
2                       0                              0   
3                       0                              0   
4                       0                              0   

   12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS TALL TUBE POSY  \
0                                    0                          0   
1                                    0                          0   
2                                    0                          0   
3                                    0                          0   
4                                    0                          0   

   12 PENCILS TALL TUBE RED RETROSPOT  12 PENCILS TALL TUBE WOODLAND  \
0                                   0                              0   
1                                   0                              0   
2                                   0                              0   
3                                   0                              0   
4                                   0                              0   

   20 DOLLY PEGS RETROSPOT  3 PIECE SPACEBOY COOKIE CUTTER SET  \
0                        0                                   0   
1                        0                                   0   
2                        0                                   0   
3                        0                                   0   
4                        0                                   0   

   3 STRIPEY MICE FELTCRAFT  36 FOIL HEART CAKE CASES  \
0                         0                         0   
1                         1                         0   
2                         0                         0   
3                         0                         0   
4                         0                         0   

   36 PENCILS TUBE WOODLAND  3D TRADITIONAL CHRISTMAS STICKERS  \
0                         0                                  0   
1                         0                                  0   
2                         0                                  0   
3                         0                                  0   
4                         0                                  0   

   4 TRADITIONAL SPINNING TOPS  50'S CHRISTMAS GIFT BAG LARGE  \
0                            0                              0   
1                            0                              0   
2                            0                              0   
3                            0                              0   
4                            0                              0   

   50'S CHRISTMAS PAPER GIFT BAG  6 GIFT TAGS 50'S CHRISTMAS  \
0                              0                           0   
1                              0                           0   
2                              0                           0   
3                              0                           0   
4                              0                           0   

   6 GIFT TAGS VINTAGE CHRISTMAS  60 CAKE CASES DOLLY GIRL DESIGN  \
0                              0                                0   
1                              0                                0   
2                              0                                0   
3                              0                                0   
4                              0                                0   

   60 TEATIME FAIRY CAKE CASES  72 SWEETHEART FAIRY CAKE CASES  \
0                            0                               0   
1                            0                               0   
2                            0                               0   
3                            0                               0   
4                            0                               0   

   A4 WALL TIDY BLUE OFFICE  ABC TREASURE BOOK BOX  ABSTRACT CIRCLE J

In [6]:
# Finding the frequent itemsets in descending order
frequent_items_portugal = apriori(portugal_basket, min_support=0.09, use_colnames=True)
frequent_items_portugal.sort_values(by=['support'],ascending=False, ignore_index=True)

support                                           itemsets
0    0.517241                                          (POSTAGE)
1    0.241379                  (RETROSPOT TEA SET CERAMIC 11 PC)
2    0.241379                          (LUNCH BAG RED RETROSPOT)
3    0.206897                     (BAKING SET 9 PIECE RETROSPOT)
4    0.206897                              (LUNCH BAG CARS BLUE)
..        ...                                                ...
105  0.103448  (PLASTERS IN TIN CIRCUS PARADE, PLASTERS IN TI...
106  0.103448  (PLASTERS IN TIN CIRCUS PARADE, PLASTERS IN TI...
107  0.103448                        (PLASTERS IN TIN STRONGMAN)
108  0.103448  (PLASTERS IN TIN SPACEBOY, PLASTERS IN TIN WOO...
109  0.103448  (LUNCH BAG CARS BLUE, JUMBO SHOPPER VINTAGE RE...

[110 rows x 2 columns]

In [7]:
# Define confidence level 
confidence_level = 0.7

# Collecting the inferred rules in a dataframe
rules_portugal = association_rules(frequent_items_portugal, metric="confidence", min_threshold=0.1)

# sort rules in descending order
rules_portugal.sort_values(by=['support'],ascending=False, ignore_index=True)

antecedents  \
0                       (BAKING SET 9 PIECE RETROSPOT)   
1                    (RETROSPOT TEA SET CERAMIC 11 PC)   
2                     (JUMBO BAG PINK VINTAGE PAISLEY)   
3                  (JUMBO SHOPPER VINTAGE RED PAISLEY)   
4                       (BAKING SET 9 PIECE RETROSPOT)   
..                                                 ...   
315  (JUMBO SHOPPER VINTAGE RED PAISLEY, JUMBO BAG ...   
316  (JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO BA...   
317                (JUMBO SHOPPER VINTAGE RED PAISLEY)   
318                          (JUMBO BAG RED RETROSPOT)   
319                   (JUMBO BAG PINK VINTAGE PAISLEY)   

                                           consequents  antecedent support  \
0                    (RETROSPOT TEA SET CERAMIC 11 PC)            0.206897   
1                       (BAKING SET 9 PIECE RETROSPOT)            0.241379   
2                  (JUMBO SHOPPER VINTAGE RED PAISLEY)            0.155172   
3                     (JUMBO BAG PINK VINTAGE PAISLEY)            0.189655   
4                                            (POSTAGE)            0.206897   
..                                                 ...                 ...   
315              (JUMBO BAG SCANDINAVIAN BLUE PAISLEY)            0.120690   
316                (JUMBO SHOPPER VINTAGE RED PAISLEY)            0.103448   
317  (JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO BA...            0.189655   
318  (JUMBO SHOPPER VINTAGE RED PAISLEY, JUMBO BAG ...            0.137931   
319  (JUMBO SHOPPER VINTAGE RED PAISLEY, LUNCH BAG ...            0.155172   

     consequent support   support  confidence      lift  leverage  conviction  \
0              0.241379  0.155172    0.750000  3.107143  0.105232    3.034483   
1              0.206897  0.155172    0.642857  3.107143  0.105232    2.220690   
2              0.189655  0.155172    1.000000  5.272727  0.125743         inf   
3              0.155172  0.155172    0.818182  5.272727  0.125743    4.646552   
4              0.517241  0.137931    0.666667  1.288889  0.030916    1.448276   
..                  ...       ...         ...       ...       ...         ...   
315            0.155172  0.103448    0.857143  5.523810  0.084721    5.913793   
316            0.189655  0.103448    1.000000  5.272727  0.083829         inf   
317            0.103448  0.103448    0.545455  5.272727  0.083829    1.972414   
318            0.137931  0.103448    0.750000  5.437500  0.084423    3.448276   
319            0.103448  0.103448    0.666667  6.444444  0.087396    2.689655   

     zhangs_metric  
0         0.855072  
1         0.893939  
2         0.959184  
3         1.000000  
4         0.282609  
..             ...  
315       0.931373  
316       0.903846  
317       1.000000  
318       0.946667  
319       1.000000  

[320 rows x 10 columns]

In [8]:
# Select rules with the chosen confidence level
selected_rules_portugal = rules_portugal[rules_portugal['confidence'] >= confidence_level]

# Get the top three rules based on support
top_three_rules_portugal = selected_rules_portugal.sort_values(by='support', ascending=False).head(3)

# Display the top three rules
display(top_three_rules_portugal[['antecedents', 'consequents', 'support', 'confidence', 'lift']])

antecedents                          consequents  \
2        (BAKING SET 9 PIECE RETROSPOT)    (RETROSPOT TEA SET CERAMIC 11 PC)   
12  (JUMBO SHOPPER VINTAGE RED PAISLEY)     (JUMBO BAG PINK VINTAGE PAISLEY)   
13     (JUMBO BAG PINK VINTAGE PAISLEY)  (JUMBO SHOPPER VINTAGE RED PAISLEY)   

     support  confidence      lift  
2   0.155172    0.750000  3.107143  
12  0.155172    0.818182  5.272727  
13  0.155172    1.000000  5.272727

The output shows the top three association rules for the Portugal dataset at the confidence_level = 0.7. The rules are displayed in a dataframe with columns for the antecedents, consequents, support, confidence, and lift.

The first rule shows that there is a strong association between the antecedent "BAKING SET 9 PIECE RETROSPOT" and the consequent "RETROSPOT TEA SET CERAMIC 11 PC". The support for this rule is 0.155, which means that 15.5% of the transactions in the dataset contain both items. The confidence is 0.75, which means that 75% of the transactions that contain the antecedent also contain the consequent. The lift is 3.11, which indicates that the association between the antecedent and consequent is 3.11 times stronger than would be expected if they were independent.

The second and third rules show that there is a strong association between the "JUMBO BAG PINK VINTAGE PAISLEY" and "JUMBO SHOPPER VINTAGE RED PAISLEY" items. The support for both rules is the same (0.155), and the confidence for the second rule is 1.0, which means that all transactions that contain the antecedent also contain the consequent. The lift for both rules is 5.27, indicating a very strong association.


In [9]:
# Define confidence levels 
confidence_level = 0.8

# Collecting the inferred rules in a dataframe
rules_portugal = association_rules(frequent_items_portugal, metric="confidence", min_threshold=0.1)

# sort rules in descending order
rules_portugal.sort_values(by=['support'],ascending=False, ignore_index=True)

# Select rules with the chosen confidence level
selected_rules_portugal = rules_portugal[rules_portugal['confidence'] >= confidence_level]

# Get the top three rules based on support
top_three_rules_portugal = selected_rules_portugal.sort_values(by='support', ascending=False).head(3)

# Display the top three rules
display(top_three_rules_portugal[['antecedents', 'consequents', 'support', 'confidence', 'lift']])

antecedents  \
12                 (JUMBO SHOPPER VINTAGE RED PAISLEY)   
13                    (JUMBO BAG PINK VINTAGE PAISLEY)   
145  (JUMBO SHOPPER VINTAGE RED PAISLEY, JUMBO BAG ...   

                             consequents   support  confidence      lift  
12      (JUMBO BAG PINK VINTAGE PAISLEY)  0.155172    0.818182  5.272727  
13   (JUMBO SHOPPER VINTAGE RED PAISLEY)  0.155172    1.000000  5.272727  
145            (LUNCH BAG RED RETROSPOT)  0.137931    0.888889  3.682540

The output shows the top three association rules for the Portugal dataset at the confidence_level = 0.8. 

The first and second rules show that there is a strong association between the "JUMBO BAG PINK VINTAGE PAISLEY" and "JUMBO SHOPPER VINTAGE RED PAISLEY" items. The support for both rules is the same (0.155), and the confidence for the first rule is 1.0, which means that all transactions that contain the antecedent also contain the consequent. The lift for both rules is 5.27, indicating a very strong association.

The third rule shows that there is also a strong association between the "JUMBO BAG PINK VINTAGE PAISLEY" and "LUNCH BAG RED VINTAGE PAISLEY" items, as well as the "JUMBO SHOPPER VINTAGE RED PAISLEY" item. The support for this rule is slightly lower (0.138), and the confidence is 1.0, indicating that all transactions that contain both antecedents also contain the consequent. The lift is the same as the previous rules (5.27), indicating a strong association

In [10]:
# Define confidence levels 
confidence_level = 0.9

# Collecting the inferred rules in a dataframe
rules_portugal = association_rules(frequent_items_portugal, metric="confidence", min_threshold=0.1)

# sort rules in descending order
rules_portugal.sort_values(by=['support'],ascending=False, ignore_index=True)

# Select rules with the chosen confidence level
selected_rules_portugal = rules_portugal[rules_portugal['confidence'] >= confidence_level]

# Get the top three rules based on support
top_three_rules_portugal = selected_rules_portugal.sort_values(by='support', ascending=False).head(3)

# Display the top three rules
display(top_three_rules_portugal[['antecedents', 'consequents', 'support', 'confidence', 'lift']])

antecedents  \
13                    (JUMBO BAG PINK VINTAGE PAISLEY)   
146  (LUNCH BAG RED RETROSPOT, JUMBO BAG PINK VINTA...   
144  (JUMBO SHOPPER VINTAGE RED PAISLEY, LUNCH BAG ...   

                             consequents   support  confidence      lift  
13   (JUMBO SHOPPER VINTAGE RED PAISLEY)  0.155172         1.0  5.272727  
146  (JUMBO SHOPPER VINTAGE RED PAISLEY)  0.137931         1.0  5.272727  
144     (JUMBO BAG PINK VINTAGE PAISLEY)  0.137931         1.0  6.444444

The output shows the top three association rules for the Portugal dataset at the confidence_level = 0.9 in descending order. 


Rule 12 indicates that when customers buy "JUMBO BAG PINK VINTAGE PAISLEY", they are highly likely (confidence = 1.0) to also buy "JUMBO SHOPPER VINTAGE RED PAISLEY", with a support of 0.155172 and a lift of 5.272727.

Rule 146 indicates that when customers buy "JUMBO SHOPPER VINTAGE RED PAISLEY" and "LUNCH BAG", they are highly likely (confidence = 1.0) to also buy "JUMBO BAG PINK VINTAGE PAISLEY", with a support of 0.137931 and a lift of 6.444444.

Rule 145 indicates that when customers buy "JUMBO BAG PINK VINTAGE PAISLEY" and "LUNCH BAG RED RETROSPOT", they are highly likely (confidence = 1.0) to also buy "JUMBO SHOPPER VINTAGE RED PAISLEY", with a support of 0.137931 and a lift of 5.272727 

confidence level 9 is the most prefered For the Portugal_Online dataset

# for Sweden_Online_Retail


In [11]:
#import sweden_oline dataset
sweden_basket = pd.read_csv('Sweden_online_retail.csv')

In [12]:
#view first 5 rows of dataset
display(sweden_basket.head())

InvoiceNo  12 PENCIL SMALL TUBE WOODLAND  \
0     538848                              0   
1     539338                              0   
2     540040                              0   
3     542428                              0   
4     542911                              0   

   12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
0                                    0                            0   
1                                    0                            0   
2                                    0                            0   
3                                    0                            0   
4                                    0                            0   

   12 PENCILS TALL TUBE WOODLAND  3 PIECE SPACEBOY COOKIE CUTTER SET  \
0                              0                                   0   
1                              0                                   0   
2                              0                                   0   
3                              0                                   0   
4                              0                                   1   

   3 RAFFIA RIBBONS 50'S CHRISTMAS  3 RAFFIA RIBBONS VINTAGE CHRISTMAS  \
0                                0                                   0   
1                                0                                   0   
2                                0                                   0   
3                                0                                   0   
4                                0                                   0   

   3 TIER CAKE TIN RED AND CREAM  3 TRADITIONAl BISCUIT CUTTERS  SET  \
0                              0                                   0   
1                              0                                   0   
2                              0                                   0   
3                              1                                   0   
4                              0                                   0   

   36 DOILIES DOLLY GIRL  36 DOILIES VINTAGE CHRISTMAS  \
0                      0                             0   
1                      1                             0   
2                      0                             0   
3                      0                             0   
4                      0                             0   

   36 FOIL HEART CAKE CASES  36 FOIL STAR CAKE CASES  \
0                         0                        0   
1                         0                        0   
2                         0                        0   
3                         0                        0   
4                         0                        0   

   4 PEAR BOTANICAL DINNER CANDLES  4 VANILLA BOTANICAL CANDLES  \
0                                0                            0   
1                                0                            0   
2                                0                            0   
3                                0                            0   
4                                0                            0   

   5 HOOK HANGER MAGIC TOADSTOOL  6 GIFT TAGS 50'S CHRISTMAS  \
0                              0                           0   
1                              1                           0   
2                              0                           0   
3                              0                           0   
4                              0                           0   

   6 RIBBONS RUSTIC CHARM  60 CAKE CASES DOLLY GIRL DESIGN  \
0                       0                                0   
1                       0                                1   
2                       0                                0   
3                       0                                0   
4                       0                                0   

   60 CAKE CASES VINTAGE CHRISTMAS  60 TEATIME FAIRY CAKE CASES  \
0                                0                            0   
1

In [13]:
# dropping the InvoiceNo column
sweden_basket = sweden_basket.drop('InvoiceNo',axis=1)

In [14]:
# Finding the frequent itemsets
frequent_items_sweden = apriori(sweden_basket, min_support=0.09, use_colnames=True)
frequent_items_sweden.sort_values(by=['support'],ascending=False, ignore_index=True)

support                                           itemsets
0   0.611111                                          (POSTAGE)
1   0.194444                 (SET OF 3 CAKE TINS PANTRY DESIGN)
2   0.166667                           (MINI PAINT SET VINTAGE)
3   0.138889                                (GUMBALL COAT RACK)
4   0.138889                           (JAM MAKING SET PRINTED)
5   0.138889           (RED TOADSTOOL LED NIGHT LIGHT, POSTAGE)
6   0.138889                       (GUMBALL COAT RACK, POSTAGE)
7   0.138889                    (RED TOADSTOOL LED NIGHT LIGHT)
8   0.138889                (WORLD WAR 2 GLIDERS ASSTD DESIGNS)
9   0.111111                  (60 CAKE CASES DOLLY GIRL DESIGN)
10  0.111111                              (WALL TIDY RETROSPOT)
11  0.111111                      (RABBIT NIGHT LIGHT, POSTAGE)
12  0.111111  (PACK OF 72 RETROSPOT CAKE CASES, PACK OF 60 S...
13  0.111111                        (WOODEN OWLS LIGHT GARLAND)
14  0.111111                           (WOODEN BOX OF DOMINOES)
15  0.111111                (ROUND SNACK BOXES SET OF 4 FRUITS)
16  0.111111               (SET OF 72 RETROSPOT PAPER  DOILIES)
17  0.111111                (SET OF 12 FAIRY CAKE BAKING CASES)
18  0.111111                  (60 CAKE CASES VINTAGE CHRISTMAS)
19  0.111111                  (RETROSPOT TEA SET CERAMIC 11 PC)
20  0.111111                               (RABBIT NIGHT LIGHT)
21  0.111111                  (PACK OF 72 RETROSPOT CAKE CASES)
22  0.111111                   (PACK OF 60 SPACEBOY CAKE CASES)
23  0.111111               (PACK OF 60 PINK PAISLEY CAKE CASES)
24  0.111111                     (WALL TIDY RETROSPOT, POSTAGE)

In [15]:
# Define confidence levels 
confidence_level = 0.7

# Collecting the inferred rules in a dataframe
rules_sweden = association_rules(frequent_items_sweden, metric="confidence", min_threshold=0.1)

# sort rules in descending order
rules_sweden.sort_values(by=['support'],ascending=False, ignore_index=True)

antecedents                        consequents  \
0                (GUMBALL COAT RACK)                          (POSTAGE)   
1                          (POSTAGE)                (GUMBALL COAT RACK)   
2    (RED TOADSTOOL LED NIGHT LIGHT)                          (POSTAGE)   
3                          (POSTAGE)    (RED TOADSTOOL LED NIGHT LIGHT)   
4  (PACK OF 72 RETROSPOT CAKE CASES)   (PACK OF 60 SPACEBOY CAKE CASES)   
5   (PACK OF 60 SPACEBOY CAKE CASES)  (PACK OF 72 RETROSPOT CAKE CASES)   
6               (RABBIT NIGHT LIGHT)                          (POSTAGE)   
7                          (POSTAGE)               (RABBIT NIGHT LIGHT)   
8              (WALL TIDY RETROSPOT)                          (POSTAGE)   
9                          (POSTAGE)              (WALL TIDY RETROSPOT)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.138889            0.611111  0.138889    1.000000  1.636364   
1            0.611111            0.138889  0.138889    0.227273  1.636364   
2            0.138889            0.611111  0.138889    1.000000  1.636364   
3            0.611111            0.138889  0.138889    0.227273  1.636364   
4            0.111111            0.111111  0.111111    1.000000  9.000000   
5            0.111111            0.111111  0.111111    1.000000  9.000000   
6            0.111111            0.611111  0.111111    1.000000  1.636364   
7            0.611111            0.111111  0.111111    0.181818  1.636364   
8            0.111111            0.611111  0.111111    1.000000  1.636364   
9            0.611111            0.111111  0.111111    0.181818  1.636364   

   leverage  conviction  zhangs_metric  
0  0.054012         inf       0.451613  
1  0.054012    1.114379       1.000000  
2  0.054012         inf       0.451613  
3  0.054012    1.114379       1.000000  
4  0.098765         inf       1.000000  
5  0.098765         inf       1.000000  
6  0.043210         inf       0.437500  
7  0.043210    1.086420       1.000000  
8  0.043210         inf       0.437500  
9  0.043210    1.086420       1.000000

In [16]:
# Select rules with the chosen confidence level
selected_rules_sweden = rules_sweden[rules_sweden['confidence'] >= confidence_level]

# Get the top three rules based on support
top_three_rules_sweden = selected_rules_sweden.sort_values(by='support', ascending=False).head(3)

# Display the top three rules
display(top_three_rules_sweden[['antecedents', 'consequents', 'support', 'confidence', 'lift']])

antecedents                       consequents  \
0                (GUMBALL COAT RACK)                         (POSTAGE)   
6    (RED TOADSTOOL LED NIGHT LIGHT)                         (POSTAGE)   
2  (PACK OF 72 RETROSPOT CAKE CASES)  (PACK OF 60 SPACEBOY CAKE CASES)   

    support  confidence      lift  
0  0.138889         1.0  1.636364  
6  0.138889         1.0  1.636364  
2  0.111111         1.0  9.000000

The output shows the top three association rules for the Sweden basket dataset with confidence level= 7

The first and second rules show that there is a strong association between the "GUMBALL COAT RACK" and "RED TOADSTOOL LED NIGHT LIGHT" items and the "POSTAGE" item. The support for both rules is the same (0.139), and the confidence is 1.0, which means that all transactions that contain the antecedent also contain the consequent. The lift for both rules is 1.64, indicating a moderate association.

The third rule shows that there is a very strong association between the "PACK OF 60 SPACEBOY CAKE CASES" and "PACK OF 72 RETROSPOT CAKE CASES" items, with a support of 0.111 and a confidence of 1.0. The lift for this rule is 9.0, indicating a very strong association.

In [17]:
# Define confidence levels 
confidence_level = 0.8

# Collecting the inferred rules in a dataframe
rules_sweden = association_rules(frequent_items_sweden, metric="confidence", min_threshold=0.1)

# sort rules in descending order
rules_sweden.sort_values(by=['support'],ascending=False, ignore_index=True)

# Select rules with the chosen confidence level
selected_rules_sweden = rules_sweden[rules_sweden['confidence'] >= confidence_level]

# Get the top three rules based on support
top_three_rules_sweden = selected_rules_sweden.sort_values(by='support', ascending=False).head(3)

# Display the top three rules
display(top_three_rules_sweden[['antecedents', 'consequents', 'support', 'confidence', 'lift']])

antecedents                       consequents  \
0                (GUMBALL COAT RACK)                         (POSTAGE)   
6    (RED TOADSTOOL LED NIGHT LIGHT)                         (POSTAGE)   
2  (PACK OF 72 RETROSPOT CAKE CASES)  (PACK OF 60 SPACEBOY CAKE CASES)   

    support  confidence      lift  
0  0.138889         1.0  1.636364  
6  0.138889         1.0  1.636364  
2  0.111111         1.0  9.000000

TThe output shows the top three association rules for the Sweden basket dataset with confidence level= 8

The first and second rules show that there is a strong association between the "GUMBALL COAT RACK" and "RED TOADSTOOL LED NIGHT LIGHT" items and the "POSTAGE" item. The support for both rules is the same (0.139), and the confidence is 1.0, which means that all transactions that contain the antecedent also contain the consequent. The lift for both rules is 1.64, indicating a moderate association.

The third rule shows that there is a very strong association between the "PACK OF 60 SPACEBOY CAKE CASES" and "PACK OF 72 RETROSPOT CAKE CASES" items, with a support of 0.111 and a confidence of 1.0. The lift for this rule is 9.0, indicating a very strong association.

In [18]:
# Define confidence levels 
confidence_level = 0.9

# Collecting the inferred rules in a dataframe
rules_sweden = association_rules(frequent_items_sweden, metric="confidence", min_threshold=0.1)

# sort rules in descending order
rules_sweden.sort_values(by=['support'],ascending=False, ignore_index=True)

# Select rules with the chosen confidence level
selected_rules_sweden = rules_sweden[rules_sweden['confidence'] >= confidence_level]

# Get the top three rules based on support
top_three_rules_sweden = selected_rules_sweden.sort_values(by='support', ascending=False).head(3)

# Display the top three rules
display(top_three_rules_sweden[['antecedents', 'consequents', 'support', 'confidence', 'lift']])

antecedents                       consequents  \
0                (GUMBALL COAT RACK)                         (POSTAGE)   
6    (RED TOADSTOOL LED NIGHT LIGHT)                         (POSTAGE)   
2  (PACK OF 72 RETROSPOT CAKE CASES)  (PACK OF 60 SPACEBOY CAKE CASES)   

    support  confidence      lift  
0  0.138889         1.0  1.636364  
6  0.138889         1.0  1.636364  
2  0.111111         1.0  9.000000

The output shows the top three association rules for the Sweden basket dataset with confidence level= 9

The first and second rules show that there is a strong association between the "GUMBALL COAT RACK" and "RED TOADSTOOL LED NIGHT LIGHT" items and the "POSTAGE" item. The support for both rules is the same (0.139), and the confidence is 1.0, which means that all transactions that contain the antecedent also contain the consequent. The lift for both rules is 1.64, indicating a moderate association.

The third rule shows that there is a very strong association between the "PACK OF 60 SPACEBOY CAKE CASES" and "PACK OF 72 RETROSPOT CAKE CASES" items, with a support of 0.111 and a confidence of 1.0. The lift for this rule is 9.0, indicating a very strong association.

All confidence levels implemented on the Sweden_Online dataset produce similar outputs

# for uk_online_retail 

In [19]:
#import uk_oline dataset
uk_basket = pd.read_csv('uk_online_retail.csv')

In [20]:
#view first 5 rows of dataset
display(uk_basket.head())

InvoiceNo  *Boombox Ipod Classic  *USB Office Mirror Ball  \
0    536365                      0                        0   
1    536366                      0                        0   
2    536367                      0                        0   
3    536368                      0                        0   
4    536369                      0                        0   

   10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
0                       0                           0   
1                       0                           0   
2                       0                           0   
3                       0                           0   
4                       0                           0   

   12 DAISY PEGS IN WOOD BOX  12 EGG HOUSE PAINTED WOOD  \
0                          0                          0   
1                          0                          0   
2                          0                          0   
3                          0                          0   
4                          0                          0   

   12 HANGING EGGS HAND PAINTED  12 IVORY ROSE PEG PLACE SETTINGS  \
0                             0                                 0   
1                             0                                 0   
2                             0                                 0   
3                             0                                 0   
4                             0                                 0   

   12 MESSAGE CARDS WITH ENVELOPES  12 PENCIL SMALL TUBE WOODLAND  \
0                                0                              0   
1                                0                              0   
2                                0                              0   
3                                0                              0   
4                                0                              0   

   12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
0                                    0                            0   
1                                    0                            0   
2                                    0                            0   
3                                    0                            0   
4                                    0                            0   

   12 PENCILS TALL TUBE POSY  12 PENCILS TALL TUBE RED RETROSPOT  \
0                          0                                   0   
1                          0                                   0   
2                          0                                   0   
3                          0                                   0   
4                          0                                   0   

   12 PENCILS TALL TUBE SKULLS  12 PENCILS TALL TUBE WOODLAND  \
0                            0                              0   
1                            0                              0   
2                            0                              0   
3                            0                              0   
4                            0                              0   

   12 PINK HEN+CHICKS IN BASKET  12 PINK ROSE PEG PLACE SETTINGS  \
0                             0                                0   
1                             0                                0   
2                             0                                0   
3                             0                                0   
4                             0                                0   

   12 RED ROSE PEG PLACE SETTINGS  15 PINK FLUFFY CHICKS IN BOX  \
0                               0                             0   
1                               0                             0   
2                               0                             0   
3                               0                             0   
4                               0                             0   

   15CM CHRISTMAS GLASS BALL 20 LIGHTS  16 PC CUTLERY SE

In [21]:
# dropping the InvoiceNo column
uk_basket = uk_basket.drop('InvoiceNo',axis=1)

In [22]:
# Finding the frequent itemsets
frequent_items_uk = apriori(uk_basket, min_support=0.02, use_colnames=True)
frequent_items_uk.sort_values(by=['support'],ascending=False, ignore_index=True)

support                                           itemsets
0    0.116034               (WHITE HANGING HEART T-LIGHT HOLDER)
1    0.103820                          (JUMBO BAG RED RETROSPOT)
2    0.090266                         (REGENCY CAKESTAND 3 TIER)
3    0.085391                                    (PARTY BUNTING)
4    0.074570                          (LUNCH BAG RED RETROSPOT)
..        ...                                                ...
364  0.020089                 (SET 3 RETROSPOT TEA,COFFEE,SUGAR)
365  0.020089                        (POPPY'S PLAYHOUSE KITCHEN)
366  0.020035  (WOODLAND CHARLOTTE BAG, STRAWBERRY CHARLOTTE ...
367  0.020035               (SET OF 60 I LOVE LONDON CAKE CASES)
368  0.020035  (JUMBO SHOPPER VINTAGE RED PAISLEY, JUMBO BAG ...

[369 rows x 2 columns]

In [23]:
# Define confidence levels 
confidence_level = 0.3

# Collecting the inferred rules in a dataframe
rules_uk = association_rules(frequent_items_uk, metric="confidence", min_threshold=0.1)

# sort rules in descending order
rules_uk.sort_values(by=['support'],ascending=False, ignore_index=True)

antecedents  \
0               (JUMBO BAG RED RETROSPOT)   
1               (JUMBO BAG PINK POLKADOT)   
2       (GREEN REGENCY TEACUP AND SAUCER)   
3       (ROSES REGENCY TEACUP AND SAUCER)   
4                (JUMBO STORAGE BAG SUKI)   
..                                    ...   
177  (WHITE HANGING HEART T-LIGHT HOLDER)   
178            (STRAWBERRY CHARLOTTE BAG)   
179              (WOODLAND CHARLOTTE BAG)   
180          (JUMBO BAG WOODLAND ANIMALS)   
181   (JUMBO SHOPPER VINTAGE RED PAISLEY)   

                             consequents  antecedent support  \
0              (JUMBO BAG PINK POLKADOT)            0.103820   
1              (JUMBO BAG RED RETROSPOT)            0.062088   
2      (ROSES REGENCY TEACUP AND SAUCER)            0.050035   
3      (GREEN REGENCY TEACUP AND SAUCER)            0.051267   
4              (JUMBO BAG RED RETROSPOT)            0.060535   
..                                   ...                 ...   
177              (HEART OF WICKER SMALL)            0.116034   
178             (WOODLAND CHARLOTTE BAG)            0.036053   
179           (STRAWBERRY CHARLOTTE BAG)            0.036589   
180  (JUMBO SHOPPER VINTAGE RED PAISLEY)            0.039374   
181         (JUMBO BAG WOODLAND ANIMALS)            0.060695   

     consequent support   support  confidence       lift  leverage  \
0              0.062088  0.042053    0.405057   6.523895  0.035607   
1              0.103820  0.042053    0.677308   6.523895  0.035607   
2              0.051267  0.037553    0.750535  14.639752  0.034988   
3              0.050035  0.037553    0.732497  14.639752  0.034988   
4              0.103820  0.037392    0.617699   5.949737  0.031108   
..                  ...       ...         ...        ...       ...   
177            0.062356  0.020089    0.173130   2.776479  0.012854   
178            0.036589  0.020035    0.555721  15.188342  0.018716   
179            0.036053  0.020035    0.547584  15.188342  0.018716   
180            0.060695  0.020035    0.508844   8.383568  0.017646   
181            0.039374  0.020035    0.330097   8.383568  0.017646   

     conviction  zhangs_metric  
0      1.576473       0.944807  
1      2.777201       0.902769  
2      3.803076       0.980765  
3      3.551237       0.982039  
4      2.344176       0.885531  
..          ...            ...  
177    1.133968       0.723819  
178    2.168481       0.969099  
179    2.130666       0.969638  
180    1.912435       0.916818  
181    1.433977       0.937629  

[182 rows x 10 columns]

In [24]:
# Select rules with the chosen confidence level
selected_rules_uk = rules_uk[rules_uk['confidence'] >= confidence_level]

# Get the top three rules based on support
top_three_rules_uk = selected_rules_uk.sort_values(by='support', ascending=False).head(3)

# Display the top three rules
display(top_three_rules_uk[['antecedents', 'consequents', 'support', 'confidence', 'lift']])

antecedents                        consequents  \
50          (JUMBO BAG PINK POLKADOT)          (JUMBO BAG RED RETROSPOT)   
51          (JUMBO BAG RED RETROSPOT)          (JUMBO BAG PINK POLKADOT)   
31  (GREEN REGENCY TEACUP AND SAUCER)  (ROSES REGENCY TEACUP AND SAUCER)   

     support  confidence       lift  
50  0.042053    0.677308   6.523895  
51  0.042053    0.405057   6.523895  
31  0.037553    0.750535  14.639752

The output shows the top three association rules for the Uk Online dataset with the confidence level of 0.3

The first and second rules show that there is a strong association between the "JUMBO BAG PINK POLKADOT" and "JUMBO BAG RED RETROSPOT" items. The support for both rules is the same (0.042), and the confidence for the first rule is 0.677, while for the second rule it is 0.405. The lift for both rules is 6.52, indicating a strong association.

The third rule shows that there is a very strong association between the "GREEN REGENCY TEACUP AND SAUCER" and "ROSES REGENCY TEACUP AND SAUCER" items, with a support of 0.0376 and a confidence of 0.751. The lift for this rule is 14.64, indicating a very strong association.

In [25]:
# Define confidence levels 
confidence_level = 0.7

# Collecting the inferred rules in a dataframe
rules_uk = association_rules(frequent_items_uk, metric="confidence", min_threshold=0.1)

# sort rules in descending order
rules_uk.sort_values(by=['support'],ascending=False, ignore_index=True)

# Select rules with the chosen confidence level
selected_rules_uk = rules_uk[rules_uk['confidence'] >= confidence_level]

# Get the top three rules based on support
top_three_rules_uk = selected_rules_uk.sort_values(by='support', ascending=False).head(3)

# Display the top three rules
display(top_three_rules_uk[['antecedents', 'consequents', 'support', 'confidence', 'lift']])

antecedents                        consequents  \
30  (ROSES REGENCY TEACUP AND SAUCER)  (GREEN REGENCY TEACUP AND SAUCER)   
31  (GREEN REGENCY TEACUP AND SAUCER)  (ROSES REGENCY TEACUP AND SAUCER)   
26   (PINK REGENCY TEACUP AND SAUCER)  (GREEN REGENCY TEACUP AND SAUCER)   

     support  confidence       lift  
30  0.037553    0.732497  14.639752  
31  0.037553    0.750535  14.639752  
26  0.030910    0.820768  16.403939

The output shows the top three association rules for the Uk Online dataset with the confidence level of 0.7

The first and second rules show that there is a very strong association between the "GREEN REGENCY TEACUP AND SAUCER" and "ROSES REGENCY TEACUP AND SAUCER" items. The support for both rules is the same (0.0376), and the confidence for the first rule is 0.751, while for the second rule it is 0.732. The lift for both rules is 14.64, indicating a very strong association.

The third rule shows that there is also a strong association between the "PINK REGENCY TEACUP AND SAUCER" and "GREEN REGENCY TEACUP AND SAUCER" items, with a support of 0.031 and a confidence of 0.821. The lift for this rule is 16.40, indicating a very strong association

In [26]:
# Define confidence levels 
confidence_level = 0.8

# Collecting the inferred rules in a dataframe
rules_uk = association_rules(frequent_items_uk, metric="confidence", min_threshold=0.1)

# sort rules in descending order
rules_uk.sort_values(by=['support'],ascending=False, ignore_index=True)

# Select rules with the chosen confidence level
selected_rules_uk = rules_uk[rules_uk['confidence'] >= confidence_level]

# Get the top three rules based on support
top_three_rules_uk = selected_rules_uk.sort_values(by='support', ascending=False).head(3)

# Display the top three rules
display(top_three_rules_uk[['antecedents', 'consequents', 'support', 'confidence', 'lift']])

antecedents  \
26                    (PINK REGENCY TEACUP AND SAUCER)   
164  (ROSES REGENCY TEACUP AND SAUCER, PINK REGENCY...   
166  (PINK REGENCY TEACUP AND SAUCER, GREEN REGENCY...   

                           consequents  support  confidence       lift  
26   (GREEN REGENCY TEACUP AND SAUCER)  0.03091    0.820768  16.403939  
164  (GREEN REGENCY TEACUP AND SAUCER)  0.02641    0.902930  18.046041  
166  (ROSES REGENCY TEACUP AND SAUCER)  0.02641    0.854419  16.666089

The output shows the top three association rules for the Uk Online dataset with the confidence level of 0.8

The first rule shows that there is a strong association between the "PINK REGENCY TEACUP AND SAUCER" and "GREEN REGENCY TEACUP AND SAUCER" items, with a support of 0.0309 and a confidence of 0.821. The lift for this rule is 16.40, indicating a very strong association.

The second rule shows that there is also a strong association between the combination of "GREEN REGENCY TEACUP AND SAUCER and PINK REGENCY TEACUP AND SAUCER" with "ROSES REGENCY TEACUP AND SAUCER" items, this has a support of 0.0264 and a confidence of 0.854. The lift for this rule is 16.67, indicating a very strong association.

The third rule shows that there is a very strong association between the combination of "ROSES REGENCY TEACUP AND SAUCER and PINK REGENCY TEACUP AND SAUCER" and "GREEN REGENCY TEACUP AND SAUCER" items, having a support of 0.0264 and a confidence of 0.903. The lift for this rule is 18.05, indicating a very strong association

confidence level 8 is the most prefered For the UK_Online dataset

(1226 words)